blockprod/src/detail/job_manager.rs

```diff
@@ -0,0 +1,358 @@
+        mut new_job_receiver: UnboundedReceiver<(
+            Id<GenBlock>,
+            UnboundedSender<()>,
+            oneshot::Sender<Result<JobKey, JobManagerError>>,
+        )>,
+        mut stop_job_receiver: UnboundedReceiver<(Option<JobKey>, oneshot::Sender<usize>)>,
+        mut shutdown_receiver: UnboundedReceiver<oneshot::Sender<usize>>,
+    ) {
+        tokio::spawn(async move {
+            let mut jobs: BTreeMap<JobKey, JobHandle> = BTreeMap::new();
+
+            loop {
+                tokio::select! {
+                    event = get_job_count_receiver.recv()
+                        => Self::get_job_count_handler(&jobs, event),
+
+                    event = chainstate_receiver.recv()
+                        => Self::chainstate_handler(&mut jobs, event),
+
+                    event = new_job_receiver.recv()
+                        => Self::new_job_handler(&mut jobs, event),
+
+                    event = stop_job_receiver.recv()
+                        => Self::stop_job_handler(&mut jobs, event),
+
+                    event = shutdown_receiver.recv()
+                        => return Self::shutdown_handler(&mut jobs, event), // Note: the return here
+                }
+            }
+        });
+    }
+
+    fn subscribe_to_chainstate(&mut self, chainstate_sender: UnboundedSender<Id<GenBlock>>) {
+        let chainstate_handle = self.chainstate_handle.clone();
+
+        tokio::spawn(async move {
+            chainstate_handle
+                .call_mut(|this| {
+                    let subscribe_func =
+                        Arc::new(
+                            move |chainstate_event: ChainstateEvent| match chainstate_event {
+                                ChainstateEvent::NewTip(block_id, _) => {
+                                    if let Err(e) = chainstate_sender.send(block_id.into()) {
+                                        log::error!(
+                                            "Chainstate subscriber failed to send new tip: {e:?}"
+                                        )
+                                    }
+                                }
+                            },
+                        );
+
+                    this.subscribe_to_events(subscribe_func);
+                })
+                .await
+        });
+    }
+
+    fn chainstate_handler(jobs: &mut BTreeMap<JobKey, JobHandle>, event: Option<Id<GenBlock>>) {
+        if let Some(new_tip_id) = event {
+            let mut jobs_to_stop: Vec<JobKey> = vec![];
+
+            for (job_key, _) in jobs.iter() {
+                if job_key.current_tip_id() != new_tip_id {
+                    jobs_to_stop.push(job_key.clone());
+                }
+            }
+
+            for job_key in jobs_to_stop {
+                if let Some(handle) = jobs.remove(&job_key) {
+                    if let Err(e) = handle.cancel_sender.send(()) {
+                        log::info!("Error sendin cancel job event: {e:?}");
```

Comment:
```suggestion
                        log::info!("Error sending cancel job event: {e:?}");
```

Reply:
I recently switched to MX Cherry Green switches, and I'm missing a lot a ',', 'o', and 'g'. This has been fixed, and am now trying to type harder to avoid misses.